### PR TITLE
types: `HardFork` improvements

### PR DESCRIPTION
**Pull request by hinto-janai**
### What
Adds docs, tests, `fn` to the `cuprate_types::HardFork` type.

### Why
Added changes used in #308.

### CODE REVIEW

File: types/src/hard_fork.rs [hunk 1]
```diff
@@ -1,6 +1,10 @@
 //! The [`HardFork`] type.
 use std::time::Duration;
 
+use strum::{
+    AsRefStr, Display, EnumCount, EnumIs, EnumString, FromRepr, IntoStaticStr, VariantArray,
+};
```
Comment (hinto-janai) [inline types/src/hard_fork.rs:6]: Hopefully pulling in `strum` is justified by replacing the manual "remember to update" code with an automated version.

File: types/src/hard_fork.rs [hunk 2]
```diff
@@ -47,58 +69,87 @@ pub enum HardFork {
...
+            assert!(COUNT <= u8::MAX as usize);
+            COUNT as u8
+        };
+
+        if version == 0 || version > COUNT_AS_U8 {
+            Err(HardForkError::HardForkUnknown)
+        } else {
+            // INVARIANT: we've proved above that `version` is in a valid range.
+            Ok(Self::VARIANTS[(version - 1) as usize])
+        }
```
Comment (hinto-janai) [inline types/src/hard_fork.rs:119]: This is also an automatic version of the old `from_version`, this now means updating `HardFork` is as simple as adding another variant.

It's a bit more complex than the `match` though...

File: types/src/hard_fork.rs [hunk 3]
```diff
@@ -109,22 +160,49 @@ impl HardFork {
...
+    ///     assert_eq!(hf.as_u8(), i as u8 + 1);
+    /// }
+    /// ```
+    pub const fn as_u8(self) -> u8 {
+        self as u8
     }
 
     /// Returns the next hard-fork.
-    pub fn next_fork(&self) -> Option<Self> {
-        Self::from_version(*self as u8 + 1).ok()
```
Comment (hinto-janai) [inline types/src/hard_fork.rs:118]: Copy is cheaper than `&`.

File: types/src/hard_fork.rs (hunk 2 above)
Comment (Boog900) [reply types/src/hard_fork.rs:119]: Couldn't we use [FromRepr](https://docs.rs/strum/0.26.3/strum/derive.FromRepr.html#) here?

File: types/src/hard_fork.rs [hunk 4]
```diff
@@ -47,58 +69,87 @@ pub enum HardFork {
...
 }
 
 impl HardFork {
+    /// The current [`HardFork`].
+    ///
+    /// ```rust
+    /// # use cuprate_types::HardFork;
+    /// assert_eq!(HardFork::CURRENT, HardFork::V16);
+    /// ```
+    pub const CURRENT: Self = Self::VARIANTS[Self::COUNT - 1];
```
Comment (Boog900) [inline types/src/hard_fork.rs:82]: Bit of nit but I think this should be called  `LATEST`  as the last HF may have not been activated, therefore it's not always the current HF.

File: types/src/hard_fork.rs [hunk 5]
```diff
@@ -109,22 +160,49 @@ impl HardFork {
...
+    ///
+    /// for hf in HardFork::VARIANTS.iter() {
+    ///     if *hf == HardFork::CURRENT {
+    ///         assert!(hf.is_current());
+    ///     } else {
+    ///         assert!(!hf.is_current());
+    ///     }
+    /// }
+    /// ```
+    pub const fn is_current(self) -> bool {
```
Comment (Boog900) [inline types/src/hard_fork.rs:205]: same here: `is_latest`?

**Review by Boog900 — APPROVED** [review]

**Pull request merged**